Static files overhaul

guardrex opened the pull request:
Fixes #35810
Fixes #35833

After Javier technical review, I'll ping Wade and Tom.
 
### Notes

* Let's combine articles. IMO, it's kind'a scatter-🧠 to have ***two*** static files articles.
* '***Look ma! No broken links!***' 😆 Let's drop the horribad INCLUDES versioning for two great outcomes ...
  * Links aren't broken in prior-version content.
  * The team and community are prevented from only updating the latest version of the article. That will still be a problem if we ever go to group-topic versioning (aka "Version-by-file"), but I don't think that's ever going to happen. If we're on the current versioning model forever, this approach stops the poor practice of only making bug fixes to the latest release content (i.e., INCLUDES files are routinely ignored for bug and enhancement updates).
* Let's move the samples into the `dotnet/AspNetCore.Docs.Samples` repo. The sample apps were placed in that repo on https://github.com/dotnet/AspNetCore.Docs.Samples/pull/282. I'll delete them from this repo on this PR ***after*** reviews to avoid cluttering the diff.
* We're still running with the separate [Blazor Static Files article](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/static-files) with Blazor-focused scenarios. Dan, I can move that content (or selected bits of it) into this article if you want. 👂


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/fundamentals/environments.md](https://github.com/dotnet/AspNetCore.Docs/blob/3111f365741233c8c177de6b7e331d7a59fbeedb/aspnetcore/blazor/fundamentals/environments.md) | [aspnetcore/blazor/fundamentals/environments](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/environments?branch=pr-en-us-35967) |
| [aspnetcore/blazor/fundamentals/static-files.md](https://github.com/dotnet/AspNetCore.Docs/blob/3111f365741233c8c177de6b7e331d7a59fbeedb/aspnetcore/blazor/fundamentals/static-files.md) | [aspnetcore/blazor/fundamentals/static-files](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/static-files?branch=pr-en-us-35967) |
| [aspnetcore/fundamentals/static-files.md](https://github.com/dotnet/AspNetCore.Docs/blob/3111f365741233c8c177de6b7e331d7a59fbeedb/aspnetcore/fundamentals/static-files.md) | [aspnetcore/fundamentals/static-files](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/static-files?branch=pr-en-us-35967) |
| [aspnetcore/migration/50-to-60-samples.md](https://github.com/dotnet/AspNetCore.Docs/blob/3111f365741233c8c177de6b7e331d7a59fbeedb/aspnetcore/migration/50-to-60-samples.md) | [aspnetcore/migration/50-to-60-samples](https://review.learn.microsoft.com/en-us/aspnet/core/migration/50-to-60-samples?branch=pr-en-us-35967) |
| [aspnetcore/migration/70-80.md](https://github.com/dotnet/AspNetCore.Docs/blob/3111f365741233c8c177de6b7e331d7a59fbeedb/aspnetcore/migration/70-80.md) | [aspnetcore/migration/70-80](https://review.learn.microsoft.com/en-us/aspnet/core/migration/70-80?branch=pr-en-us-35967) |
| [aspnetcore/toc.yml](https://github.com/dotnet/AspNetCore.Docs/blob/3111f365741233c8c177de6b7e331d7a59fbeedb/aspnetcore/toc.yml) | [aspnetcore/toc](https://review.learn.microsoft.com/en-us/aspnet/core/toc?branch=pr-en-us-35967) |


<!-- PREVIEW-TABLE-END -->

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1104 @@
+* An asset is served once until the file changes or the browser clears its cache. The [`ETag`](https://developer.mozilla.org/docs/Web/HTTP/Headers/ETag) and [Last-Modified](https://developer.mozilla.org/docs/Web/HTTP/Headers/Last-Modified) headers are set.
+* The browser is prevented from using stale assets after an app is updated. The [`Last-Modified`](https://developer.mozilla.org/docs/Web/HTTP/Headers/Last-Modified) header is set.
+* The appropriate [caching headers](https://developer.mozilla.org/docs/Web/HTTP/Headers/Cache-Control) are set on the response.
+* [Caching Middleware](xref:performance/caching/middleware) is used.
```
The Caching middleware isn't actually a feature of MapStaticAssets. It's a separate independent feature.

```suggestion
```

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1104 @@
+
+<xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A> serves static files, but it doesn't provide the same level of optimization as <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A>. <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A> is optimized for serving assets that the app has  ...
+
+Map Static Assets provides the following benefits that aren't available when calling <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A>:
```
Another benefit of MapStaticAssets is that it is based on endpoint routing, so it works with other endpoint aware features like authorization.

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1104 @@
+
+During development, the framework overrides cache headers to prevent browsers from caching static files. This helps ensure that the latest version of files are used when files change, avoiding issues with stale content. In production, the correct cache headers are set, allowing browsers to cache static assets as expected.
+
+To disable this behavior, set `EnableStaticAssetsDevelopmentCaching` to `false` in the Development environment's app setting file (`appsettings.Development.json`).
```
I think this property might also be specific to MapStaticAssets, but I'm not 100% sure.

guardrex commented on aspnetcore/fundamentals/static-files.md:
````diff
@@ -1,17 +1,1104 @@
+});
+```
+
+When developing a server-side Blazor app and testing locally, see <xref:blazor/fundamentals/static-files#static-files-in-non-development-environments>.
````
There were behavioral deltas when the article was set up. As a general principal, only deltas (or Blazor-focused scenarios) were covered by the Blazor-specific articles. I'll double-check and get back to you, but it might be Tuesday, given that I'm trying to get the *Environments* article worked wrapped up before the holiday weekend.

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1104 @@
+
+:::moniker range=">= aspnetcore-6.0"
+
+*The guidance in this section applies to Razor Pages and MVC apps. For guidance that applies to Blazor Web Apps, see <xref:blazor/fundamentals/static-files#serve-files-from-multiple-locations>.*
```
Isn't the guidance the same? Can we remove this duplication?

guardrex replied:
That's what I meant about deltas. It wasn't the same IIRC, but I'll have to look. I'll check and get back to you ... probably Tuesday morning.

danroth27 commented on aspnetcore/fundamentals/static-files.md:
````diff
@@ -1,17 +1,1104 @@
+app.MapStaticAssets();
+```
+
+## Static file authorization
````
I'm a bit confused about what this section is trying to communicate. It sounds like it's trying to explain how you can add authorization checks for your static files. But then the first thing it shows is enabling anonymous access to static web assets. Shouldn't this section start off with showing that static web assets can have authorization policies applied just like any other endpoint?

The rest of the section seems to be about workarounds for enabling authorization when using the static files middleware. This content is certainly important for folks still using the static files middleware, but it dominates this section in a way that I don't think we want.

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1104 @@
+* Integrates the information gathered about static web assets during the build and publish process with a runtime library that processes this information to optimize file serving to the browser.
+* Are routing endpoint conventions that optimize the delivery of static assets in an app. It's designed to work with all UI frameworks, including Blazor, Razor Pages, and MVC.
+
+## Static Files Middleware (`UseStaticFiles`)
```
I'd like to avoid presenting the reader with a choice of how to handle static files right at the start of this article. Instead, I think the article should start off focused on basic static file handling using MapStaticAssets, almost as if we were documenting static file handling in ASP.NET Core for the first time and MapStaticAssets is the only thing available. After we've covered all the basic stuff, we can then cover scenarios that aren't supported by MapStaticAssets and that require the static file middleware.

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1104 @@
+
+:::moniker range=">= aspnetcore-9.0"
+
+## Map Static Assets (`MapStaticAssets`)
```
Can we start with some basics?:

- To enable static file handling in ASP.NET Core, call `MapStaticAssets`
- You put any files you want to be served as static files in your wwwroot folder (also called the web root)
- By default, only files in the wwwroot folder will be addressable, so you don't need to worry about the rest of your code.
- Only files with specific file extensions mapped to supported media types will be treated as static web assets.
- Your static web assets will get discovered at build time and optimized using content-based finger printing and precompession.
- At runtime, the discovered static web assets will be exposed as endpoints and will have appropriate HTTP headers applied, like caching headers, content type, etc.
- Note that you can also serve static web assets from referenced projects and packages. Link to related content.

danroth27 reviewed (changes requested):
Great start! I'd like to simplify the beginning of the static files doc as much as possible and gradually build up to more complicated and legacy scenarios. Dealing with static files is pretty common, so the basics need to come across as simple and straightforward. I also want to consolidate more of the general purpose static web asset content in the top-level static files doc instead of having a bunch of it duplicated in the Blazor. The Blazor static files doc really should only cover Blazor specific stuff, like the framework specific static web assets that it adds, the helpers it provides for referencing static web assets (`Assets[]`), etc.

guardrex commented:
@danroth27 ... Ready for another look 👀. Notes on the changes ...

## Some discussion above

There are two points of discussion above ☝️.

## Focusing on Map Static Assets

> presenting the reader with a choice of how to handle static files right at the start of this article ... MapStaticAssets is the only thing available. After we've covered all the basic stuff, we can then cover scenarios that aren't supported by MapStaticAssets and that require the static file middleware.

For .NET 9 or later, Static Files Middleware is pushed down into the article immediately before where it's first needed.

I also added a < .NET 9 bit for introductory content on Static File Middleware, which I should've placed earlier.

## Duplicate content

> instead of having a bunch of it duplicated in the Blazor

"*A bunch*"?? 😖😆 ... That was never the intention for Blazor's fundamentals coverage. Only in limited cases is there duplication. In ancient Blazor times 👴, I checked scenarios from the main doc set's articles to see where deltas required special coverage. That's what ended up in the Blazor *Fundamentals* node articles. The current Blazor fundamentals coverage only contains rather specific guidance for Blazor, and we can still move stuff (or all of it) over. I'm still wondering why any fundamental coverage (or host/deploy ... or security) is covered in the Blazor node, given Blazor's preeminent position. MVC/RP nodes never carried such content the way that we're carrying it today.

Map Static Assets coverage for .NET 9's release was slow to materialize in the main doc set. I had to document it fully in the Blazor node to get it covered for release, but it was rectified later. That content was placed in the separate *Map Static Assets* article and dropped from Blazor's article. I think we agree that ***one*** static files article makes sense.

Except for whatever we do with the *Serve static files from multiple locations* section (probably needs Javier's input, links are above ☝️), there's only one other section to address ... [*File mappings and static file options* section](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/static-files?view=aspnetcore-9.0#file-mappings-and-static-file-options). There was a major delta in versions prior to .NET 8. I'm updating the coverage now to drop the content in the Blazor article from .NET 8 forward. We're keeping the Blazor coverage for < .NET 8 due to the change in the way the Blazor script is served.

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1150 @@
+
+By default, static files are stored within the project's [web root](xref:fundamentals/index#web-root) directory. The default directory is `{CONTENT ROOT}/wwwroot`, where the `{CONTENT ROOT}` placeholder is the app's [content root](xref:fundamentals/index#content-root). Only files in the wwwroot folder will be addressable, so you don't need to worry about the rest of your code.
+
+Only files with specific file extensions mapped to supported media types are treated as static web assets.
```
@javiercn Is there more we should say about the default media type mapping for static web assets and how it can be configured?

javiercn replied:
Yes, I included a section about that in my original technical description. it should be covered here if it's not covered later in the document. At leas a note/link here to where it can be configured later.

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1150 @@
+
+Only files with specific file extensions mapped to supported media types are treated as static web assets.
+
+Static web assets are discovered at build time and optimized using content-based [fingerprinting](https://wikipedia.org/wiki/Fingerprint_(computing)) to prevent the reuse of old file and [compressed](/aspnet/core/performance/response-compression) to reduce asset delivery time. [The optimizations don't include minification.]
```
```suggestion
Static web assets are discovered at build time and optimized using content-based [fingerprinting](https://wikipedia.org/wiki/Fingerprint_(computing)) to prevent the reuse of old files.  Assets are also [compressed](/aspnet/core/performance/response-compression) to reduce asset delivery time.
```

danroth27 commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1150 @@
+
+Map Static Assets provides the following benefits:
+
+* Build-time compression for all the assets in the app, including JavaScript (JS) and stylesheets but excluding image and font assets that are already compressed. [Gzip](https://tools.ietf.org/html/rfc1952) (`Content-Encoding: gz`) compression is used during development. Gzip with [Brotli](https://tools.ietf.org/html/rfc7932) (`Content-Encoding: br`) compression is used during publish.
```
```suggestion
* Build-time compression for all the assets in the app, including JavaScript (JS) and stylesheets but excluding image and font assets that are already compressed. [Gzip](https://tools.ietf.org/html/rfc1952) (`Content-Encoding: gz`) compression is used during development. Gzip and [Brotli](https://tools.ietf.org/html/rfc7932) (`Content-Encoding: br`) compression are both used during publish.
```

javiercn commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1150 @@
+`DisableBuildCompression` | Disables compression during build.
+`CompressDiscoveredAssetsDuringBuild` | Compresses discovered assets during build.
+`BrotliCompressionLevel` | Compression level for the Brotli algorithm.
+`LinkAlternativeRepresentationsToOriginalResource` | How to link compressed assets to original resources.
```
We should remove this property, I missed it when I created the list, it should not be here.

guardrex replied:
```suggestion
```

javiercn commented on aspnetcore/fundamentals/static-files.md:
```diff
@@ -1,17 +1,1150 @@
+
+Item group | Description | Metadata
+--- | ---
+`StaticWebAssetContentTypeMapping` | Maps file patterns to content types and cache headers for endpoints. | Pattern, Cache
```
This should be covered under a section about "adding new content type mappings at build time"

guardrex replied:
Ok ... I have one more pass to get the sample app out of here. It's in the samples repo now. I didn't do it yet because I didn't want to clutter the diff. I'll take care of it by EOD.

javiercn reviewed (approved):
Few comments, otherwise looks good.
